spring/spring4/spring4-common/src/main/java/org/infinispan/spring/provider/CacheDelegate.java

```diff
@@ -71,13 +77,54 @@ public ValueWrapper get(final Object key) {
 
    @Override
    public <T> T get(Object key, Callable<T> valueLoader) {
-      return (T) nativeCache.computeIfAbsent(key, keyToBeInserted -> {
-         try {
-            return valueLoader.call();
-         } catch (Exception e) {
-            throw ValueRetrievalExceptionResolver.throwValueRetrievalException(key, valueLoader, e);
+      StampedLock lock = null;
+      long stamp = -1;
+      // In order to maintain parallel access to this get methods we need to maintain a map of locks. Each lock
+      // is associated with a key. The lock exists by the time this method executes value loader passed by the argument.
+      // The algoritms performs 3 checks:
+      // 1 - without synchronisation, it checks if the value is there. If it's not - 2nd step.
+      // 2 - with read lock. If the value is still missing, 3rd step.
+      // 3 - we upgrade from read to write lock and read the value from loader.
+      // Finally, we remove the lock from the lock map.
+      try {
+         T value = (T) nativeCache.get(key);
+         if (value == null) {
+            lock = synchronousGetLocks.computeIfAbsent(key, k -> new StampedLock());
+            stamp = lock.readLock();
+            value = (T) nativeCache.get(key);
+            if (value == null) {
+               while (true) {
+                  long upgradedLockStamp = lock.tryConvertToWriteLock(stamp);
+                  if (upgradedLockStamp != 0) {
+                     stamp = upgradedLockStamp;
+                     if ((value = (T) nativeCache.get(key)) == null) {
+                        try {
+                           T newValue = valueLoader.call();
+                           // we can't use computeIfAbsent here since in distributed embedded scenario we would
+                           // send a lambda to other nodes. This is the behavior we want to avoid.
+                           value = (T) nativeCache.putIfAbsent(key, newValue);
+                           if (value == null) {
+                              value = newValue;
+                           }
+                        } catch (Exception e) {
+                           throw ValueRetrievalExceptionResolver.throwValueRetrievalException(key, valueLoader, e);
+                        }
+                     }
+                     break;
+                  } else {
+                     lock.unlockRead(stamp);
+                     stamp = lock.writeLock();
+                  }
+               }
+            }
+         }
+         return value;
+      } finally {
+         if (lock != null) {
+            lock.unlock(stamp);
+            synchronousGetLocks.remove(key);
```

Comment:
This does not sound safe. Suppose that:
T1 creates and acquires lock
T2 retrieves the lock and blocks
T1 unlocks the lock and removes it
T2 can proceed, enters the critical section
T3 comes and creates another lock, does not get blocked by T2...

Reply:
> T1 creates and acquires lock

Assuming T1 acquires the Read Lock at line 93.

> T2 retrieves the lock and blocks

T2 successfully upgraded its lock at line 97 and now blocks at line 102.

> T1 unlocks the lock and removes it

Impossible. T1 tries to upgrade its Read Lock to Write Lock (just as a reminder, Write Locks are exclusive) by executing `lock.tryConvertToWriteLock(stamp);` (line 97). This returns `0` since T2 is holding the write lock and sitting at line 102. T1 enters the `else` block and unlocks the Read Lock. Now T1 also blocks but on line 116, it's trying to acquire the Write Lock exclusively but it's waiting till T2 releases it (`writeLock()` is a blocking method).

Unless I misunderstood this, the situation you showed seems to be impossible in reality.

Reply:
No, I meant a situation where:

> T1 acquires read lock, upgrades to write lock on 97
> T2 tries to acquire a read lock and blocks on 93
> T1 unlocks and removes the lock
> T2 unblocks and has read-lock acquired
> T3 comes and does not see the same instance of lock as T2

The T3 can now get the computed value (if it did not expire/be evicted in the meantime) or whatever; the thing is that you have a concept of critical section here and might use different instances of locks in concurrent threads visiting this section.

Reply:
hmmm good point. I believe we need to hire some visiting thread counting mechanism here. The idea is to remove the lock only if the current thread is the last one visiting this section.

Reply:
@slaskawi You're only using the read lock after `cache.get(key)` already returned `null` once, so I think you can skip it and use a `DefaultLockManager` to manage the locking.

Reply:
Ok, so how about simplifying this code and synchronizing on the key? This way it will be super easy to read and will probably do what we want. The only downside is that the key we got in an argument might be already locked by the client code. However this is the thing we could just document. 

Looking for up/down vote from @rvansa and @danberindei.

Reply:
@danberindei This is the implementation of a shared Cache Delegate. It can be used in both remote and embedded scenario. Therefore I can not use `DefaultLockManager`.

Reply:
@rvansa @danberindei I drastically simplified locking here. Please take another look at it.